### ISSUE TITLE
Replace by parameter- is there currently a better way?

### ISSUE THREAD

**Issue by mxdmedia** [briebug/ngrx-auto-entity#80]
Back with another question/potential feature request.  Lets say I have two dependent models:
```typescript
export class Parent {
  @Key id?: string;
  name: string;
}

export class Child {
  @Key id?: string;
  name: string;
  parentId: string;
}
```
In my API, I have an endpoint, `api/parent/[parentId]/children` which allows CRUD updates on `Child` elements in bulk.  E.g.- if I `PUT` to `api/parent/[parentId]/children` it will update any children of `[parentId]` in the request that already exist, create any new children (without an id) and delete any children that were not passed in the request.

What is the best way to handle the response (especially the delete part) if both `Parent` and `Child` are managed via auto-entity, and there are multiple parents?  I am thinking my effect needs to do something like:

```typescript
export const updateParentChildren = createAction(
  '[Child] Update Parent Children',
  props<{ payload: {parent: Parent, children: Child[] }>()
);

// ... in my effect
 loadJobDate$ = createEffect(
    () =>
      this.actions$.pipe(
        ofType(updateParentChildren),
        map(action => action.payload),
        exhaustMap(request =>
          this.parentService.updateChildren(request).pipe(
            withLatestFrom(this.store$.pipe(select(selectChildAll))), // Get all children
            map(([response, allChildren]) => {  // response is array of Children to replace Parent's children with
              // Get children to delete.  They must 1) have same parent as request,
              // and 2) not be in the repsonse.
              const deleteChildren = allChildren
                .filter(child => child.parentId===request.parent.id)
                .filter(oldChild => !(response.find(respChild => respChild.id === oldChild.id)));
              this.store$.dispatch(new DeleteManySuccess(Child, deleteChildren)); // delete Parent's children that were not in response
              this.store$.dispatch(new LoadManySuccess(Child, response)); // Add parents new children
            })
          )
        )
      ),
    { dispatch: false }
  );
```
This should work, however I am wondering if there is a cleaner way of doing this using either the current auto-entity tools, or using any of the future Entity related stuff that has been discussed.  As one who works with a lot of related data like this, I can see something like a `ReplaceManyRelated` action as being useful.  It would replace a subset of objects, based on some lookup.  Something like:
```typescript
    this.store$.dispatch(new ReplaceManyRelated(Child, {parent: parent.id}, response));
```
Where `{parentId: parent.id}` would indicate the related field to use (e.g.- `Child.parentId`) and the value to select from `parent.id`.  Thoughts?

**Comment by paquettealice** [comment]
Hey!

Although I can't put them down right this moment (I will do so later tonight or in the next few days), I have a few ideas on how to simplify things for the short term, after which we can see what could be done as far as a long term solution is concerned (i.e. new feature).

Also, thanks for your continued feedback and use of the library!

**Comment by jrista** [comment]
@mxdmedia I wanted to introduce you to Alice, who has recently become a contributor to NgRx Auto-Entity. She has some ideas that may help. 

She also recently just did some work to add a feature to the library (coming soon) that may help you as well.

**Comment by mxdmedia** [comment]
@jrista Thank you for the introduction.

@paquettealice Nice meeting you. I look forward to hearing your thoughts on this (when you have time), as well as seeing the coming features of auto-entity.  So far auto-entity has been an excellent tool for me, that has helped to scratch many of my ngrx "itches," and I am excited to see where it goes!

**Comment by paquettealice** [comment]
@mxdmedia Nice to meet you!

I looked at this a bit more, but unfortunately, and despite having ideas on how we might simplify things, most of them felt like a rather marginal improvement and not worth pointing out, while the more promising ones were a bit too involved for me to really explore. My personal bandwidth is being stretched to the limit lately, so I've had some trouble finding time for this project.

That said, I've had to deal with 'synchronizing' arrays of child entities in the past, and it definitely is quite tedious to do. Because of this, I think it's worth looking into a 'long term' solution that would address this particular pain point. I might make a proposal in the coming weeks, but no promises.

Finally, sorry I couldn't be more helpful! Perhaps you'll find some inspiration after one of the new releases? 0.3.0, [which is currently in PR](https://github.com/briebug/ngrx-auto-entity/pull/79), introduces correlated actions -- those are useful when chaining multiple actions, which is what you're doing (in a way) in your example. 

And, though I don't yet know the down-and-dirty specifics, Jon also has been working on some pretty big stuff, so look forward to that :)

**Comment by mxdmedia** [comment]
@paquettealice, my apologies, I just saw that you had responded to this.. a month ago.  Thank you for the insight, and thoughts.  No worries on not being able to provide an immediate solution.  As you pointed out, this kind of parent-child entity relationship can be tricky, and I have yet to find an "easy" solution with any react/ngrx style state management libraries/add-ons.  For me, that is the holy grail, as my code uses a lot of these sorts of relationships.  It is great seeing `auto-entity` grow, and it looks like there are things in the pipeline that will help in these sorts of operations.

It looks like 0.3.x has now been released.  I am in the middle of a soft launch of my big project (which uses `auto-entity`) so I wont be upgrading from 0.2.8 immediately (if it ain't broke, dont fix it), but I'll definitely be looking into it in the coming weeks.  Looking through the release notes, it appears that with 0.3.1 there are no breaking changes, as it doesn't require the `uuid` library.  Is this correct?

**Comment by jrista** [comment]
@mxdmedia Still thinking about this one. I think we are at the fine line between what belongs in the library, vs. what needs to be the responsibility of the developer.

I'm not quite sure on this one. Partly because I don't think we've come up with any solution yet that really feels like it fits. But we are still thinking about it, and if we do find a solution that fits, we'll let you know.

**Comment by jrista** [comment]
@mxdmedia 

I think this particular use case may need to remain a responsibility of the developer. That said, we just released [0.4.1](https://github.com/briebug/ngrx-auto-entity/releases/tag/0.4.1), which includes the ability to delete things just by their keys, rather than requiring entities, which could help. 

I also think you can do some things to improve your usage of NgRx in general, and avoid some common anti-patterns. Your example above uses `store.dispatch` within effects, which is strongly discouraged.  You can actually avoid that, with some additional custom actions and effects. 

Auto-entity now supports selection of entities. So if you need to work with a particular parent, you can select it. You can then build additional effects around that selection:

```typescript
const {initialState: initialParentState, selectors: {
    selectCurrentEntity: selectedParent,
    selectCurrentEntityKey: selectedParentId
}} = buildState(Parent);
const {initialState: initialChildState, selectors: {selectAll: allChildren}} = buildState(Child);

export const mapToChildrenOfParent = (parentId: string, children: Child[]) => 
    children.map(child => child.parentId === parentId);

export const childrenOfSelectedParent = createSelector(
    selectedParentId,
    allChildren,
    mapToChildrenOfParent
);
```

I also highly recommend dropping the use of payload in your actions, as that is now also an anti-pattern and can just make action data harder to use:

```typescript
export const updateParentChildren = createAction(
  '[Child] Update Parent Children',
  props<{parent: Parent, children: Child[]}>()
);
```

This is the recommended approach with new NgRx 8 actions, and allows you to simply destructure the data directly out of the action:

```typescript
ofType(updateParentChildren),
exhaustMap(({parent, children}) => ...)
```

You may only need the children in your action, though, if you use selection:

```typescript
export const updateSelectedParentsChildren = createAction(
  '[Child] Update Parent Children',
  props<{children: Child[]}>()
);

export const oldChildrenReplaced = createAction(
    '[Child] Old Children Replaced'
    props<{newChildren: Child[], oldChildren: Child[]}>
);

// Somewhere in your code, select the parent:
this.parents.select(parent);
this.parents.selectByKey(parentId);

// In effects:
 loadJobDate$ = createEffect(
    () => this.actions$.pipe(
        ofType(updateSelectedParentsChildren),
        withLatestFrom(this.store$.select(selectedParent)),
        map([{children}, parent]) => ({ parent, children })),
        exhaustMap(request =>
            this.parentService.updateChildren(request).pipe(
                withLatestFrom(this.store$.select(childrenOfSelectedParent))),
                map(([newChildren, oldChildren]) => oldChildrenReplaced({newChildren, oldChildren))
            )
         )
    ),
    { dispatch: false }
  );

  deleteOldChildren$ = createEffect(
    () => this.actions$.pipe(
        ofType(oldChildrenReplaced),
        map(({newChildren, oldChildren}) => oldChildren.filter(oldChild =>
            !(newChildren.find(newChild => newChild.id === oldChild.id)
        )),
        map(deleteChildren => new DeleteManySuccess(Child, deleteChildren))
    )
  );

  replaceOldChildren$ = createEffect(
    () => this.actions$.pipe(
        ofType(oldChildrenReplaced),
        map(({newChildren}) => new LoadManySuccess(Child, newChildren))
    )
  );
```

This avoids calling store.dispatch within effects, and leverages effects for what they are designed for: to perform side effects and transform actions.
